extension/js/common/core/crypto/pgp/openpgp-key.ts

```diff
@@ -269,114 +244,99 @@ export class OpenPGPKey {
-      fullyEncrypted: keyWithoutWeakPackets.isPublic() ? false /* public keys are never encrypted */ : keyWithoutWeakPackets.isFullyEncrypted(),
-      isPublic: keyWithoutWeakPackets.isPublic(),
+      fullyDecrypted,
+      fullyEncrypted,
+      isPublic: !keyWithoutWeakPackets.isPrivate(),
       isPrivate: keyWithoutWeakPackets.isPrivate(),
       algo: {
         algorithm: algoInfo.algorithm,
         bits: algoInfo.bits,
-        curve: (algoInfo as unknown as { curve: string | undefined }).curve,
+        curve: algoInfo.curve,
         algorithmId: opgp.enums.publicKey[algoInfo.algorithm],
       },
       revoked: keyWithoutWeakPackets.revocationSignatures.length > 0,
     } as Key);
-    (key as Key & { internal: OpenPGP.key.Key }).internal = keyWithoutWeakPackets;
-    (key as Key & { rawKey: OpenPGP.key.Key }).rawKey = opgpKey;
-    (key as Key & { rawArmored: string }).rawArmored = opgpKey.armor();
+    const keyWithPrivateFields = key as KeyWithPrivateFields;
+    keyWithPrivateFields.internal = keyWithoutWeakPackets;
+    keyWithPrivateFields.rawKey = opgpKey;
+    keyWithPrivateFields.rawArmored = opgpKey.armor();
     return key;
   };
 
   /**
    * Returns signed data if detached=false, armored
    * Returns signature if detached=true, armored
    */
-  public static sign = async (signingPrivate: Key, data: string, detached = false): Promise<string> => {
-    const signingPrv = OpenPGPKey.extractExternalLibraryObjFromKey(signingPrivate);
-    const message = opgp.cleartext.fromText(data);
-    const signRes = await opgp.sign({ message, armor: true, privateKeys: [signingPrv], detached });
-    if (detached) {
-      if (typeof signRes.signature !== 'string') {
-        throw new Error('signRes.signature unexpectedly not a string when creating detached signature');
-      }
-      return signRes.signature;
-    }
-    return await opgp.stream.readToEnd((signRes as OpenPGP.SignArmorResult).data);
+  public static sign = async (signingPrivate: Key, text: string, detached = false): Promise<string> => {
+    const signingPrv = (await OpenPGPKey.extractExternalLibraryObjFromKey(signingPrivate)) as OpenPGP.PrivateKey; // todo: throw?
+    const message = await opgp.createMessage({ text });
+    const signRes = await opgp.sign({ message, format: 'armored', signingKeys: [signingPrv], detached });
+    return signRes;
   };
 
   public static revoke = async (key: Key): Promise<string | undefined> => {
-    let prv = OpenPGPKey.extractExternalLibraryObjFromKey(key);
+    let prv = await OpenPGPKey.extractExternalLibraryObjFromKey(key);
+    if (!prv.isPrivate()) {
+      return undefined; // todo: or throw?
+    }
     if (!(await prv.isRevoked())) {
-      prv = await prv.revoke({});
+      const keypair = await opgp.revokeKey({ key: prv, format: 'object' });
+      // todo: save this data into `key`?
```

Comment:
I'm not entirely sure now what is it that we are actually doing. Did we revoke the key, or did we create a revocation certificate that can be saved and later uploaded to public key servers?

If the latter (that's what I thought), people sometimes do that as a form of forward-safety, save a revocation cert somewhere more accessible than the private key itself. not that any FlowCrypt usecases rely on that.

I think the method naming is ambiguous, UI says `revocation certificate` but this method says `revoke` when it should say `newRevocationCertificate` or similar, in which case your note would not apply.

Reply:
Right, I renamed this method to `getOrCreateRevocationCertificate()` as it either extracts the revocation certificate from the key (both public and private) if it's already revoked, or creates a new revocation certificate (if the key is private, otherwise throws an exception).
I also added `applyRevocationCertificate()` method and a unit test.

Reply:
sounds good